Fixes #31857 - Move login-page caption to footer

Ron-Lavi opened the pull request:
The caption text, which is derived from the login footer text setting
may push the login pane to the bottom of the page:

![Screenshot (55)](https://user-images.githubusercontent.com/26363699/107628508-63671d00-6c69-11eb-86ef-3bf5435de2db.png)

The small refactor moves it back to the footer where it used to live in the older login page:

![Screenshot (56)](https://user-images.githubusercontent.com/26363699/107646942-3fb0d080-6c83-11eb-8bc2-1b3ce4c4402f.png)

![Screenshot (57)](https://user-images.githubusercontent.com/26363699/107646954-43dcee00-6c83-11eb-9694-652f150b2ba0.png)




<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

Ron-Lavi commented:
Ack from @MariSvirik about the design :D

Ron-Lavi commented:
rebased, @MariaAga @amirfefer can you take a look please? thanks

MariaAga commented on webpack/assets/javascripts/react_app/components/LoginPage/components/LoginPageFooter.js:
```diff
@@ -0,0 +1,40 @@
+          variant="link"
+          onClick={() => setShowLess(!showLess)}
+        >
+          {__('Read')} {showLess ? __('more') : __('less')}
```
Can this be `__('Read more')` and `__('Read less')` in case other languages have different structure for this?

MariaAga commented on webpack/assets/javascripts/react_app/components/LoginPage/LoginPage.scss:
```diff
@@ -14,6 +14,25 @@ $background_image: linear-gradient(
+      font-size: inherit;
+      padding: 0;
+      margin-left: 10px;
+      color: #52b0fb;
```
Can you use a specif value from pf4 variables? or set a new color in `webpack/assets/javascripts/react_app/common/colors.scss` ?

MariaAga reviewed (changes requested):
Tested and it looks good!
I'm thinking maybe all the Login components should be in the same folder? but it's out of this PRs scope probably :)
left a few comment

amirfefer commented on webpack/assets/javascripts/react_app/components/LoginPage/components/LoginPageFooter.js:
```diff
@@ -0,0 +1,40 @@
+import { translate as __ } from '../../../common/I18n';
+
+const LoginPageFooter = ({ caption }) => {
+  const [showLess, setShowLess] = React.useState(true);
```
nit: could you please import `useState` directly?

amirfefer commented on webpack/assets/javascripts/react_app/components/LoginPage/components/LoginPageFooter.js:
```diff
@@ -0,0 +1,40 @@
+
+  if (caption.length > 500) {
+    loginCaption = (
+      <React.Fragment>
```
nit: we can use the short fragment syntax -  `<>` 
```js
<>
  <ComponentA />
  <ComponentB />
</>
```

amirfefer commented on webpack/assets/javascripts/react_app/components/LoginPage/components/LoginPageFooter.js:
```diff
@@ -0,0 +1,40 @@
+        <Button
+          className="btn-expand"
+          variant="link"
+          onClick={() => setShowLess(!showLess)}
```
state setters functions are async (and are not activated immediately), therefore we can't trust the current state value.
This approach is safer: 
```js
setShowLess(prevState => !prevState)
```

amirfefer reviewed (changes requested):
Thanks @laviro 👍 
I left a few comments, please also add a test for `LoginPageFooter`

ekohl commented:
> The caption text, which is derived from the login footer text setting
> may push the login pane to the bottom of the page:

You can include this in the commit message as well. That makes the lives of maintainers easier.

Ron-Lavi commented:
Thanks @amirfefer @ekohl! I think I addressed your suggestions :)

ofedoren commented on webpack/assets/javascripts/react_app/components/LoginPage/LoginPageFooter.js:
```diff
@@ -0,0 +1,36 @@
+  if (!caption) return null;
+  let loginCaption = <p>{caption}</p>;
+
+  if (caption.length > 500) {
```
Just nit that can be ignored :D

Isn't it better to store such value somewhere in a `constants.js`? Maybe even in a file of general constants, so it can be reused somewhere else?

ofedoren commented on webpack/assets/javascripts/react_app/components/LoginPage/LoginPageFooter.js:
```diff
@@ -0,0 +1,36 @@
+  if (caption.length > 500) {
+    loginCaption = (
+      <>
+        <p>{showLess ? `${caption.substring(0, 500)}...` : caption}</p>
```
^^ here for example

ekohl replied:
:+1: for having constants. Back in university my programming professor would fail us on this.

Ron-Lavi replied:
Sure, I will add it :)
Do you have a good name for it?
how about `LONG_TEXT_BREAK_CHAR`?

ofedoren replied:
Yeah, the naming is the hardest part :D

What about `MAX_LONG_TEXT`?

Ron-Lavi replied:
feels a bit too generic, @Mariaaga @amirfefer @ekohl any ideas? :D

ekohl replied:
Honestly I'm not even convinced the show more/show less is needed at all. There appears to be plenty of space so why do we need to do it? Users can scroll if needed. I'd vote for removal and always show the full text.

Ron-Lavi replied:
Good question @ekohl, I will speak with @marisvirik and @rohoover about it,
let's see what they think...

some of the text can be missed if the user didn't scroll, and a button might be more clear... though I am not against your idea :)

MariSvirik replied:
@laviro @ekohl If there are no law requirements to show the entire text at once, I would stick with Read more option even though we have enough space to show it. 
1. Keeping the entire text visible would make the Login page look crowded and busy. 
2. As a user I would be confused if it's important for me to read the text in order not to miss any essential information for my next actions. Fistly, as a user, I would find the text non-essential because of its tiny font size and location in the footer. On the other hand, it's too long to just ignore it. So I believe it could create some kind of psychological distress on the page where we want users just to focus on logging in.

In conclusion, my two cents are that we don't want to overwhelm the users on the very first page with a long text area with minimum font size, which is not essential for their next actions. So keep Read more truncation.

But we can still work on the content of the actual text with the Content team and create a text variant when the Read more want be necessary and the text won't be too long.

ekohl replied:
I don't understand this. By default there is no text at all. There is a setting that is the login text but the admin controls this value. That means that any text that is there should be deemed important to an admin, right? The admin may even be annoyed by the Read More button.

I remain unconvinced that there's positive value in a Read More button for users and for us as maintainers there's a negative value.

Ron-Lavi replied:
the `Read more` will be visible only if the caption length is more than 500 chars long,
I am not sure how long users usually write those texts, maybe we should make some limit to this setting,
so we will know for sure that the users will write a max of 1000 chars for example, so we can adjust the page,
with 1000 chars I don't think we need the Read More/Less.

btw Twitter uses 280 char limit :P

MariSvirik replied:
@ekohl I apologize, I got confused about the main goal of the text (its creation, etc.), I thought it's just some kind of a prefabricated legal disclaimer. Thank you @laviro for an explanation. My proposal is to keep it this way until we get some feedback from customers so our decisions are based on data, not our assumptions.

ekohl replied:
I think we should drop `Read More` until there's a need for it. We never had it. That would keep this PR scoped to just "move" and not "move + add max char limit". Combining would also not be clear for users who read the changelog and thus be a surprise for upgrading users.

Ron-Lavi replied:
Sounds good, if we'll need it, we can add it in another PR..

ronlavi2412 committed:
Fixes #31857 - Move login-page caption to footer

The caption text, which is derived from the login footer text setting,
may push the login pane to the bottom of the page

ekohl reviewed (commented):
From a change perspective :+1: but I can't really review the Javascript code.

amirfefer reviewed (approved):
Thanks @laviro 👍
